activesupport/lib/active_support/testing/assertionless_tests.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Testing
+    # Allows to configure the behavior when the test case has no assertions in it.
+    # This is helpful in detecting broken tests that do not perform intended assertions.
+    module AssertionlessTests # :nodoc:
+      def self.prepended(klass)
+        klass.singleton_class.attr_reader :assertionless_tests_behavior
+        klass.extend ClassMethods
+        klass.assertionless_tests_behavior = :ignore
+      end
+
+      module ClassMethods
+        def assertionless_tests_behavior=(behavior)
+          message = "Test is missing assertions"
+
+          @assertionless_tests_behavior =
+            case behavior
+            when :ignore
+              nil
+            when :log
+              logger =
+                if defined?(Rails.logger) && Rails.logger
+                  Rails.logger
+                else
+                  require "active_support/logger"
+                  ActiveSupport::Logger.new($stderr)
+                end
+
+              -> { logger.warn(message) }
+            when :raise
+              -> { raise Minitest::Assertion, message }
+            else
+              raise ArgumentError, "assertionless_tests_behavior must be one of :ignore, :log, or :raise."
+            end
+        end
+      end
+
+      def after_teardown
+        super
+
+        return if skipped? || error?
+
+        if assertions == 0
+          ActiveSupport::TestCase.assertionless_tests_behavior&.call
```

Comment:
I think people find it helpful if we could pass `name` (if that's a correct getter for the test name) so we could include it in both the error message or the log message

Reply:
Added a name and also a file/line combination to be easier detectable, as in the example for spec style given by @Earlopain.

Reply:
This is pretty great. Minitest prints relative paths by default in recent versions though, I have created a patch that does that here as well. It uses a custom assertion class to control the location:

```patch
diff --git a/activesupport/lib/active_support/testing/assertionless_tests.rb b/activesupport/lib/active_support/testing/assertionless_tests.rb
index 862b0f3d45..25d1d07c32 100644
--- a/activesupport/lib/active_support/testing/assertionless_tests.rb
+++ b/activesupport/lib/active_support/testing/assertionless_tests.rb
@@ -11,6 +11,15 @@ def self.prepended(klass)
         klass.assertionless_tests_behavior = :ignore
       end
 
+      class NoAssertions < Minitest::Assertion
+        attr_reader :location
+
+        def initialize(message, test_location)
+          @location = test_location
+          super(message)
+        end
+      end
+
       module ClassMethods
         def assertionless_tests_behavior=(behavior)
           @assertionless_tests_behavior =
@@ -26,9 +35,9 @@ def assertionless_tests_behavior=(behavior)
                   ActiveSupport::Logger.new($stderr)
                 end
 
-              ->(message) { logger.warn(message) }
+              ->(message, test_location, test_name) { logger.warn("#{message}: `#{test_name}` #{test_location}") }
             when :raise
-              ->(message) { raise Minitest::Assertion, message }
+              ->(message, test_location, *) { raise NoAssertions.new(message, test_location) }
             else
               raise ArgumentError, "assertionless_tests_behavior must be one of :ignore, :log, or :raise."
             end
@@ -42,8 +51,8 @@ def after_teardown
 
         if assertions == 0 && (behavior = ActiveSupport::TestCase.assertionless_tests_behavior)
           file, line = self.class.instance_method(name).source_location
-          message = "Test is missing assertions: `#{name}` #{file}:#{line}"
-          behavior.call(message)
+          test_location = "#{file}:#{line}"
+          behavior.call("Test is missing assertions", test_location, name)
         end
       end
     end
diff --git a/activesupport/test/testing/assertionless_tests_test.rb b/activesupport/test/testing/assertionless_tests_test.rb
index eb7b21285a..90cfbf99df 100644
--- a/activesupport/test/testing/assertionless_tests_test.rb
+++ b/activesupport/test/testing/assertionless_tests_test.rb
@@ -51,10 +51,12 @@ class AssertionlessTestsRaise < ActiveSupport::TestCase
     module AfterTeardown
       def after_teardown
         with_assertionless_tests_behavior(:raise) do
-          assert_raise(Minitest::Assertion,
-                       match: /Test is missing assertions: `test_without_assertions` .+assertionless_tests_test\.rb:\d+/) do
+          e = assert_raise(Minitest::Assertion, match: "Test is missing assertions") do
             super
           end
+          report = Minitest::Result.from(self)
+          report.failures << e
+          assert_match(/\[test\/testing\/assertionless_tests_test.rb:\d+]/, report.to_s)
         end
       end
     end
```

As far as I can tell this is all done through public api. It's pretty much only the `location` method, which is [documented](https://docs.seattlerb.org/minitest/Minitest/Assertion.html#method-i-location). Sample output:

```
Failure:
AssertionlessTestsTest::AssertionlessTestsRaise#test_without_assertions [test/testing/assertionless_tests_test.rb:7]:
Test is missing assertions
```

Reply:
Thanks for the suggestion but I am not sure these complications are needed and I am not comfortable overriding `location` method, even though it is a public method. 

Let's see what others will say about this.